src/main/java/build/buildfarm/worker/AbstractPipelineStage.java

```diff
@@ -0,0 +1,108 @@
+// Copyright 2017 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buildfarm.worker;
+
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ArrayBlockingQueue;
+
+abstract class AbstractPipelineStage implements PipelineStage {
+  private final PipelineStage output;
+  private final PipelineStage error;
+  private final BlockingQueue<OperationContext> queue;
+
+  private PipelineStage input;
+  protected boolean claimed;
+  private boolean closed;
+
+  AbstractPipelineStage(PipelineStage output, PipelineStage error) {
+    this.output = output;
+    this.error = error;
+
+    queue = new ArrayBlockingQueue<>(1);
+    input = null;
+    claimed = false;
+    closed = false;
+  }
+
+  public void setInput(PipelineStage input) {
+    this.input = input;
+  }
+
+  @Override
+  public void run() {
+    while (!output.isClosed() || claimed) {
+      iterate();
+    }
+    close();
+  }
+
+  protected void iterate() {
+    OperationContext operationContext;
+    try {
+      operationContext = take();
+      OperationContext nextOperationContext = tick(operationContext);
+      if (nextOperationContext != null && output.claim()) {
+        output.offer(nextOperationContext);
+      } else {
+        error.offer(operationContext);
+      }
+    } catch(InterruptedException ex) {
+      return;
+    } finally {
+      release();
+    }
+    after(operationContext);
+  }
+
+  protected abstract OperationContext tick(OperationContext operationContext);
+
+  protected void after(OperationContext operationContext) { }
+
+  public OperationContext take() throws InterruptedException {
+    return queue.take();
+  }
+
+  public void offer(OperationContext operationContext) {
+    queue.offer(operationContext);
+  }
+
+  @Override
+  public synchronized boolean claim() throws InterruptedException {
+    while (!closed && claimed) {
+      wait();
+    }
+    if (closed) {
+      return false;
+    }
+    claimed = true;
+    return true;
+  }
+
+  @Override
+  public synchronized void release() {
+    claimed = false;
+    notify();
```

Comment:
This will not notify all threads, it will only notify one (the notifyAll method satisfies the all case). The notify is unspecified, which makes it a this-call, on the stage object, which makes that object itself a 'per-stage lock object' (that also happens to be the 'per-stage object')